http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/service/AppServiceInterp.scala

```diff
@@ -1167,6 +1167,8 @@ final class LeoAppServiceInterp[F[_]: Parallel](config: AppServiceConfig,
           .flatMap(c => ListAppResponse.fromCluster(c, Config.proxyConfig.proxyUrlBase, labels))
           .toVector
     }
+    // If no visible clusters, check if user should be able to use this endpoint at all
+    _ <- if (res.isEmpty) authProvider.checkUserEnabled(userInfo) else F.unit
```

Comment:
Hmm... I know that in all cases except for when the list is empty sam handles TOS logic, so this is technically correct. However, I think its a bit more readable to check this up front and fail early rather than query the db first. (since failure is related to user info, which we have before DB calls)

Reply:
I'm totally down with that - are we okay with the overhead of always making this additional request? This would add one additional round trip to Sam for every list request from an enabled user. @jdcanas @lucymcnatt

Reply:
Yeah... I guess I'm not too sure how often this is happening. If we think it will frequently rejected, we should same time on DB call. If we want to be cautious and think its an edge case, we can leave it like this and save some us the lookup time. Could leave it like this, and add a metric call to count if this happens often so we can reverse it? Performance-wise it could be worth doing it lazily at the end like this...

Reply:
I don't imagine this would be happening very often because it only applies to users who have registered but have not yet accepted ToS.

Reply:
would it make sense to add a follow-up to make this the only Sam check, ie remove it from wherever its currently hitting Sam because that check will have already been taken care of with this call?

Reply:
I'd prefer to leave it as is, since the current design does not actually expose any info - the security "loophole" only allows unauthenticated users to deduce that there are no results for their query; if they make a request which would actually return resources Sam will block them as intended. Adding overhead to authenticated requests seems like a poor trade for closing that hole.

Reply:
We still need the other Sam checks (which ask if the user has permission to perform the relevant action on the resource). The check we're adding here just asks if they have permission to even use terra basically.

Reply:
if we decide to leave as is, it would be helpful if the comment was a little clearer on why this check is happening where it is, something like `If no visible clusters, Sam wasn't called to verify users permission to perform action, need to check if user should be able to use this endpoint at all`

Reply:
> would it make sense to add a follow-up to make this the only Sam check, ie remove it from wherever its currently hitting Sam because that check will have already been taken care of with this call?

I think what you're asking is whether we should add this additional check to Sam to all calls. Because Leo's existing requests to Sam happen in the context of a resource + an action - so "for disk D, user U, action A, can U perform A on D?"

The check I'm adding has only the user and the action, and without a resource there isn't enough info to satisfy the above permission model. So we'll still need to call Sam for all the resources involved in the request, and the end result works out to calling Sam an extra time for good requests (your recommendation) vs calling Sam a bunch more times for bad requests (my code). 

Definitely upsides and downsides to both! For example, a malicious requester could perhaps flood these endpoints with requests forcing lots of pointless lookups, since we're actually doing expensive resource checks for unauthenticated users. But I imagine there are plenty of ways to DDoS Terra and it might be out of scope to try to close those vulnerabilities.